tests/unit/s2n_server_extensions_test.c

```diff
@@ -456,6 +458,48 @@ int main(int argc, char **argv)
             EXPECT_SUCCESS(s2n_connection_free(conn));
         }
 
+        /* Test that some TLS1.3 extensions (like PSK) not sent on a HRR request */
+        {
+            s2n_extension_type_id psk_extension_id = 0;
+            EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(s2n_server_psk_extension.iana_value, &psk_extension_id));
+
+            struct s2n_psk psk = { 0 };
+
+            for (size_t is_hrr = 0; is_hrr < 2; is_hrr++) {
+                struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+                EXPECT_NOT_NULL(conn);
+                EXPECT_SUCCESS(s2n_connection_allow_all_response_extensions(conn));
+                conn->actual_protocol_version = S2N_TLS13;
+                struct s2n_stuffer *io_stuffer = &conn->handshake.io;
+
+                /* Setup required for PSK extension */
+                conn->psk_params.chosen_psk = &psk;
+                S2N_CBIT_CLR(conn->extension_requests_sent, psk_extension_id);
+                EXPECT_TRUE(s2n_server_psk_extension.should_send(conn));
+
+                /* Setup required for other server extensions */
+                conn->secure.cipher_suite = &s2n_tls13_aes_128_gcm_sha256;
+                conn->secure.server_ecc_evp_params.negotiated_curve = &s2n_ecc_curve_secp256r1;
+                conn->secure.client_ecc_evp_params[0].negotiated_curve = &s2n_ecc_curve_secp256r1;
+                EXPECT_SUCCESS(s2n_ecc_evp_generate_ephemeral_key(&conn->secure.client_ecc_evp_params[0]));
+
+                if (is_hrr) {
+                    EXPECT_SUCCESS(s2n_set_connection_hello_retry_flags(conn));
+                }
+
+                EXPECT_SUCCESS(s2n_server_extensions_send(conn, io_stuffer));
+
+                s2n_parsed_extensions_list parsed_extensions = { 0 };
+                EXPECT_SUCCESS(s2n_extension_list_parse(io_stuffer, &parsed_extensions));
+
+                bool psk_extension_sent = (parsed_extensions.parsed_extensions[psk_extension_id].extension_type
+                             == s2n_server_psk_extension.iana_value);
+                EXPECT_NOT_EQUAL(psk_extension_sent, is_hrr);
```

Comment:
Are we sure this comparison of bool and int won't fail in any compiler?

Reply:
Yes, C defines implicit conversions.